coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/transformers/UnionSqlCallTransformer.java

```diff
@@ -0,0 +1,165 @@
+import org.apache.calcite.sql.SqlCallBinding;
+import org.apache.calcite.sql.SqlDataTypeSpec;
+import org.apache.calcite.sql.SqlIdentifier;
+import org.apache.calcite.sql.SqlKind;
+import org.apache.calcite.sql.SqlNode;
+import org.apache.calcite.sql.SqlNodeList;
+import org.apache.calcite.sql.SqlSelect;
+import org.apache.calcite.sql.SqlTypeNameSpec;
+import org.apache.calcite.sql.fun.SqlCastFunction;
+import org.apache.calcite.sql.fun.SqlStdOperatorTable;
+import org.apache.calcite.sql.parser.SqlParserPos;
+import org.apache.calcite.sql.type.SqlTypeName;
+import org.apache.calcite.sql.validate.SqlValidator;
+import org.apache.calcite.sql.validate.SqlValidatorScope;
+
+import com.linkedin.coral.common.transformers.SqlCallTransformer;
+import com.linkedin.coral.common.utils.TypeDerivationUtil;
+
+import static org.apache.calcite.rel.rel2sql.SqlImplementor.POS;
+import static org.apache.calcite.sql.parser.SqlParserPos.ZERO;
+
+
+/**
+ * This transformer class is used to adapt expressions in
+ * set statements (e.g.: `UNION`, `INTERSECT`, `MINUS`)
+ * in case that the branches of the set statement contain fields in
+ * char family which have different types.
+ * The `char` fields which are differing from the expected `varchar` output
+ * of the set statement will be adapted through an explicit `CAST` to the `varchar` type.
+ *
+ * @see <a href="https://github.com/trinodb/trino/issues/9031">Change char/varchar coercion in Trino</a>
+ */
+public class UnionSqlCallTransformer extends SqlCallTransformer {
+
+  public UnionSqlCallTransformer(TypeDerivationUtil typeDerivationUtil) {
+    super(typeDerivationUtil);
+  }
+
+  @Override
+  protected boolean condition(SqlCall sqlCall) {
+    return sqlCall.getOperator().kind == SqlKind.UNION || sqlCall.getOperator().kind == SqlKind.INTERSECT
+        || sqlCall.getOperator().kind == SqlKind.MINUS;
+  }
+
+  @Override
+  protected SqlCall transform(SqlCall sqlCall) {
+    List<SqlNode> operandsList = sqlCall.getOperandList();
+    List<List<Optional<RelDataType>>> columnsTypesLists = new ArrayList<>();
+    for (SqlNode operand : operandsList) {
+      if (operand.getKind() == SqlKind.SELECT && ((SqlSelect) operand).getSelectList() != null) {
+        SqlSelect select = (SqlSelect) operand;
+        List<SqlNode> selectNodes = select.getSelectList().getList();
+        if (columnsTypesLists.isEmpty()) {
+          for (int i = 0; i < selectNodes.size(); i++) {
+            columnsTypesLists.add(new ArrayList<>());
+          }
+        }
+
+        for (int i = 0; i < selectNodes.size(); i++) {
+          SqlNode sqlNode = selectNodes.get(i);
+          if (sqlNode.getKind() == SqlKind.IDENTIFIER && ((SqlIdentifier) sqlNode).isStar()) {
+            // The type derivation of * on the SqlNode layer is not supported now.
+            return sqlCall;
+          }
+
+          Optional<RelDataType> selectNodeDataType;
+          try {
+            selectNodeDataType = Optional.of(deriveRelDatatype(sqlNode));
```

Comment:
@aastha25 in some cases while running the `HiveToTrinoConverterTest#testViews` I stumbled on `RuntimeException` issued from `deriveRelDataType(SqlNode)`.

```
java.lang.RuntimeException: Failed to derive the RelDataType for SqlNode: CAST(row("b"."b1") as row("b1" varchar)) AS `b` with topSqlNode: SELECT `tablef`.`a`, `generic_project`(`tablef`.`b`, 'b', 'struct<b1:string>') AS `b`
FROM `test`.`tablef` AS `tablef`
	at com.linkedin.coral.common.utils.TypeDerivationUtil.getRelDataType(TypeDerivationUtil.java:87)
	at com.linkedin.coral.common.transformers.SqlCallTransformer.deriveRelDatatype(SqlCallTransformer.java:68)
```

Is this expected?
I currently workaround this limitation by just working with column types for which the type can actually be derived.

Also side note - Why throwing the very generic `RuntimeException` in `TypeDerivationUtil` instead of specialized exceptions?

Reply:
@findinpath were you able to isolate the test case which is leading to the failure?

Reply:
Tests failing:

- `fuzzy_union_view_double_branch_evolved_different`
- `fuzzy_union_view_more_than_two_branches_evolved`
- `fuzzy_union_view_map_with_struct_value_evolved`
- `fuzzy_union_view_array_with_struct_value_evolved`
- `fuzzy_union_view_deeply_nested_struct_evolved`
- `fuzzy_union_view_deeply_nested_complex_struct_evolved`

Reply:
@ljfgem Can we fix the generic type derivation method and also throw the underlying error?

Reply:
Sure, will take a look.